### ISSUE TITLE
verbs: fails with multiple threads and mr_cache enabled

### ISSUE THREAD

**Issue by mshiryaev** [ofiwg/libfabric#4802]
Hello,

Fails happen when multiple threads work on dedicated EPs in parallel and with mr_cache enabled.
Attached reproducer: [ofi_test.zip](https://github.com/ofiwg/libfabric/files/2838794/ofi_test.zip)

To reproduce execute:

on server node
FI_VERBS_MR_CACHE_ENABLE=1 FI_PROVIDER=verbs ./ofi_test 1 server_hostname port

on client node
FI_VERBS_MR_CACHE_ENABLE=1 FI_PROVIDER=verbs ./ofi_test 0 server_hostname port

Error output from test:
...
RegisterBuffers: ret -259, error Unknown error -259, fi_mr_reg:

Test description:
each side opens EP_COUNT endpoints and starts the same count of threads. Each thread works only with dedicated EP and execute the following pattern multiple times - register buffers, do rma operation, deregister buffers.

Without mr_cache (FI_VERBS_MR_CACHE_ENABLE=0) test works well.

**Comment by mshiryaev** [comment]
@shefty , @a-ilango - did you have a chance to look at this issue?

**Comment by shefty** [comment]
I have not yet.  I will try to take a look at it in a bit to see where the locking is missing.

**Comment by shefty** [comment]
Can you try with the changes in pull request #4867 ?

**Comment by mshiryaev** [comment]
Seems this fixed issue, at least few sequential launches work well. Thanks!
But I noticed significant increase for test timings as now all threads are serialized around single domain lock. Do you have plans to make mr_cache as per-EP object?

**Comment by shefty** [comment]
There are no plans, but I guess we could configure the cache this way.  That could result in registering the same memory at each endpoint.  It depends on whether the threads have dedicated regions or not.

If we either move the locking into the cache or modify the cache APIs, we could convert the spinlock into reader-writer locks.  We could also look at releasing the lock when registering memory (since that's a lengthy operation), which might allow for better progress when one thread has a cache miss.

I'll go ahead with the current fix, since we need it for correctness, and see how we can improve parallelism.

**Comment by mshiryaev** [comment]
@shefty , thanks for fix and information. The ticket can be closed.

**Comment by shefty** [comment]
Closing, but opening new issue related to optimizing the locking
